diff --git a/src/arch/i386/prefix/romprefix.S b/src/arch/i386/prefix/romprefix.S
index 35d037e..9fa7d6d 100644
--- a/src/arch/i386/prefix/romprefix.S
+++ b/src/arch/i386/prefix/romprefix.S
@@ -422,33 +422,6 @@ no_pmm:
 	xorw	%di, %di
 	cs rep	movsb
 
-	/* Prompt for POST-time shell */
-	movw	$init_message_prompt, %si
-	xorw	%di, %di
-	call	print_message
-	movw	$prodstr, %si
-	call	print_message
-	movw	$init_message_dots, %si
-	call	print_message
-	/* Wait for Ctrl-B */
-	movw	$0xff02, %bx
-	call	wait_for_key
-	/* Clear prompt */
-	pushf
-	xorw	%di, %di
-	call	print_kill_line
-	movw	$init_message_done, %si
-	call	print_message
-	popf
-	jnz	2f
-	/* Ctrl-B was pressed: invoke iPXE.  The keypress will be
-	 * picked up by the initial shell prompt, and we will drop
-	 * into a shell.
-	 */
-	movl	$0xa0000, %ebp	/* Inhibit relocation during POST */
-	pushw	%cs
-	call	exec
-2:
 	/* Restore registers */
 	popw	%gs
 	popw	%fs
@@ -581,9 +554,6 @@ init_message_pmm:
 init_message_int19:
 	.asciz	" INT19"
 	.size	init_message_int19, . - init_message_int19
-init_message_prompt:
-	.asciz	"\nPress Ctrl-B to configure "
-	.size	init_message_prompt, . - init_message_prompt
 init_message_dots:
 	.asciz	"..."
 	.size	init_message_dots, . - init_message_dots
